tests/python_compiler/conftest.py

```
@@ -47,13 +48,24 @@ def _run_filecheck_impl(program_str, pipeline=()):
     ctx = Context(allow_unregistered=True)
     xdsl_module = QuantumParser(ctx, program_str, extra_dialects=(test.Test,)).parse_module()
 
+    if roundtrip:
+        # Print generic format
+        stream = StringIO()
+        Printer(stream=stream, print_generic_format=True).print_op(xdsl_module)
+        ctx = Context(allow_unregistered=True)
+        xdsl_module = QuantumParser(
+            ctx, stream.getvalue(), extra_dialects=(test.Test,)
+        ).parse_module()
+
     pipeline = PassPipeline(pipeline)
     pipeline.apply(ctx, xdsl_module)
 
+    stream = StringIO()
+    Printer(stream).print_op(xdsl_module)
```

```suggestion
    text_to_check: str
    try:
        if roundtrip:
            # Print generic format
            stream = StringIO()
            Printer(stream=stream, print_generic_format=True).print_op(xdsl_module)
            ctx = Context(allow_unregistered=True)
            xdsl_module = QuantumParser(
                ctx, stream.getvalue(), extra_dialects=(test.Test,)
            ).parse_module()

        if verify:
            xdsl_module.verify()
        pipeline = PassPipeline(pipeline)
        pipeline.apply(ctx, xdsl_module)
        if verify:
            xdsl_module.verify()

        stream = StringIO()
        Printer(stream).print_op(xdsl_module)
        text_to_check = stream.getvalue()
    except Exception as err:
        text_to_check = str(err)
```

---

Why is `text_to_check` the error string if an exception is raised? I don't see why a try-except block is used, because if an error is raised then we should just let it be raised right?

---

The usual pattern for testing intentionally raised errors is to use the `pytest.raises` context manager. I would say that is a better pattern just because it is more consistent with our usual practices. Happy to discuss further; the more questions like this we can resolve, the better :)